src/stdlib/ListUtils.scillib

```diff
@@ -361,56 +321,41 @@ let list_zip_with : forall 'A. forall 'B. forall 'C. ('A -> 'B -> 'C) ->
   tfun 'B =>
   tfun 'C =>
   fun (f : 'A -> 'B -> 'C) =>
-  fun (m1 : List 'A) =>
-  fun (m2 : List 'B) =>
-    let list_zip_helper =
-      tfun 'A =>
-      tfun 'B =>
-      tfun 'C =>
-      fun (g : 'A -> 'B -> 'C) =>
-      fun (l1 : List 'A) =>
-      fun (l2 : List 'B) =>
-        let folder = @list_foldl 'A (Pair (List 'C) (List 'B)) in
-        let nilb = Nil {'B} in
-        let nilc = Nil {'C} in
-        let init = Pair {(List 'C) (List 'B)} nilc l2 in
-        let iter =
-          fun (z : Pair (List 'C) (List 'B)) =>
-          fun (h : 'A) =>
-            match z with
-            | Pair r b =>
-              (* Get b's head, pair it with h and add to r. *)
-              let header = @list_head 'B in
-              let tailer = @list_tail 'B in
-              let bhead = header b in
-              match bhead with
-              | Some bel =>
-                let newp = g h bel in
-                let newp_concat = Cons {'C} newp r in
-                let btail = tailer b in
-                let newb =
-                  match btail with
-                  | Some t =>
-                    t
-                  | None =>
-                    nilb
-                  end
-                in
-                Pair {(List 'C) (List 'B)} newp_concat newb
-              | None =>
-                z
-              end
-            end
-          in
-            folder iter init l1
-    in
-      let zipper = @list_zip_helper 'A 'B 'C in
-      let res = zipper f m1 m2 in
-      match res with
-      | Pair x y =>
-        let reverser = @list_reverse 'C in
-          reverser x
+  fun (l1: List 'A) =>
+  fun (l2: List 'B) =>
+  (* Fail when l1 empty, empty l2 case done by foldk *)
+  let zip_step = fun (current_new : Pair (List 'C) (List 'A)) =>
+  fun (y : 'B) =>
+    match current_new with
+    | Pair current new =>
+      match new with
+      | Cons x xs => let next = f x y in
+                     let ad = Cons {'C} next current in
+                     let ad_rem = Pair { (List 'C) (List 'A) } ad xs in
+                     Some {(Pair (List 'C) (List 'A))} ad_rem
+      | Nil => None {(Pair (List 'C) (List 'A))}
       end
+    end in
+  let foldWhile = @list_foldl_while 'B (Pair (List 'C) (List 'A)) in
```

Comment:
I went over `ListUtils.scilla` and changed `folder` to `foldl` / `foldr` / `foldk`. IMO, it makes it easier to understand what is going on there, because one doesn't have to keep in mind the kind of fold used.

Reply:
Let me mark this as a resolved conversation, but feel free to reopen.